### ISSUE TITLE
Method return type problem when dealing with arrays

### ISSUE THREAD

**Issue by Asterix112** [lukeautry/tsoa#398]
If you define a request handler and want to return an array of some type you will face a type naming issue in the OpenAPI specification when writing
```
public testGet(): T[] {}
```
instead of
```
public testGet(): Array<T> {}
```

If you then try to generate an API client using the Swagger codegen executable you will face the issue of return types like this (Dart):
```
static List<T[]> listFromJson(List<dynamic> json) {}
```
Expected behavior:
```
static List<T> listFromJson(List<dynamic> json) {}
```

**Comment by dgreene1** [comment]
Before I try to reproduce this, please include
* the full controller file that uses this model
* the interface model definition
* the Swagger YAML that is produced
* the Swagger YAML that you wish was produced

**Comment by dgreene1** [comment]
@Asterix112 it's been 5 days and we still haven't received any steps to reproduce this. Please see above. Due to that we'll close this for housekeeping purposes. However, if you are able to submit the information above we can re-open the issue and try to solve it. Thank you.

**Comment by Asterix112** [comment]
@dgreene1 Sorry for my late response. Because of project protection purposes i cannot send the whole controller to you. I could send you everything that is directly in touch with the problem we are facing (e.g. controller function including annotations and the controller class itself, also including annotations and of course the interfaces)

Controller:
```
export interface BaseSchool {
    sid: string;
    name: string;
    city: string;
    theme: SchoolTheme;
    authRequired: boolean;
}

export interface GenericResponse<T> {
    state: ResponseState;
    data: T;
    error?: string;
}

@Route('')
@injectable()
export class DefaultController extends Controller {
    @inject(LoggingServiceImpl)
    private logging: LoggingServiceImpl;

    @inject(ControlsServiceImpl)
    private controls: ControlsServiceImpl;

    constructor() {
        super();
    }

    @Get('schools')
    public async schoolsGet(@Request() request: ExpressRequest): Promise<GenericResponse<BaseSchool[]>> {
        this.logging.info(`User: ${JSON.stringify(request.user)}`);

        this.logging.info("Retrieving schools from Controls backend...");
        const schools = await this.controls.getAllSchools(false);

        this.logging.info("Sending response...");
        return {
            data: <BaseSchool[]>schools.filter(e => !e.hidden).map(e => {
                return {
                    sid: e.sid,
                    name: e.name,
                    city: e.city,
                    theme: e.theme,
                    authRequired: e.auth.required,
                };
            }),
            state: ResponseState.SUCCEEDED,
        };
    }
}
```

Swagger:
```
{
  "components": {
    "examples": {},
    "headers": {},
    "parameters": {},
    "requestBodies": {},
    "responses": {},
    "schemas": {
      "ResponseState": {
        "enum": [
          "succeeded",
          "failed"
        ],
        "type": "string"
      },
      "BaseSchool": {
        "properties": {
          "sid": {
            "type": "string",
            "description": "* BaseSchool ID **"
          },
          "name": {
            "type": "string",
            "description": "* Full name of the school **"
          },
          "city": {
            "type": "string",
            "description": "* City where the school is located in **"
          },
          "theme": {
            "$ref": "#/components/schemas/SchoolTheme",
            "description": "* Color theme **"
          },
          "authRequired": {
            "type": "boolean",
            "description": "* User based authentication is required to login (else: username: _anonymous, password: _anonymous) **"
          }
        },
        "required": [
          "sid",
          "name",
          "city",
          "theme",
          "authRequired"
        ],
        "type": "object"
      },
      "GenericResponseBaseSchool[]": {
        "properties": {
          "state": {
            "$ref": "#/components/schemas/ResponseState"
          },
          "data": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/BaseSchool"
            }
          },
          "error": {
            "type": "string",
            "nullable": true
          }
        },
        "required": [
          "state",
          "data"
        ],
        "type": "object"
      }
    },
    "securitySchemes": {
      "sia": {
        "type": "oauth2",
        "clientId": "",
        "clientSecret": "",
        "tokenUrl": "https://schulinfoapp-v3b.apis.cedgetec.com/oauth/token",
        "flow": "password",
        "scopes": {
          "sia": "For accessing the API."
        }
      }
    }
  },
  "info": {
    "title": "<...>",
    "version": "<,,,>",
    "description": "<...>",
    "license": {
      "name": "<...>"
    }
  },
  "openapi": "3.0.0",
  "paths": {
    "/schools": {
      "get": {
        "operationId": "SchoolsGet",
        "produces": [
          "application/json"
        ],
        "responses": {
          "200": {
            "description": "Ok",
            "schema": {
              "$ref": "#/components/schemas/GenericResponseBaseSchool[]"
            }
          }
        },
        "security": [],
        "parameters": []
      }
    }
  },
  "servers": [
    {
      "url": "<...>"
    }
  ]
}
```

Are you fine with this and then able to figure out what the issue might be?

**Comment by dgreene1** [comment]
The swagger looks accurate to me based off of the TypeScript types you’re showing. What is wrong with the Schema that was generated?

You mentioned dart in the original description. Are you saying that the problem is in the dart code that you are generating from the swagger documentation? If so, then it’s not an issue with tsoa.

**Comment by lazylazyllama** [comment]
What he is talking about is:

"GenericResponseBaseSchool[]" should be "GenericResponseBaseSchool" ... without the "[]".

They are redundant and only cause problems. Example given: You cannot name a function "GenericResponseBaseSchool[]" because the brackets are not allowed in function names.

Also if you write Array<GenericResponseBaseSchool> the output is correct and you get "GenericResponseBaseSchool" and not something like "Array<GenericResponseBaseSchool>"

**Comment by Asterix112** [comment]
Please reopen ASAP

**Comment by dgreene1** [comment]
Okay, now that this issue has been made clear, we can reopen it.

For anyone looking to take on this PR for us, this is the acceptance criteria:

* given the interfaces described above, the $ref name that is produced should be `”GenericResponseBaseSchool"` not `”GenericResponseBaseSchool[]”`

**Comment by dgreene1** [comment]
@asterix112 I reopened it and updated the acceptance criteria above to clarify that the produced name should not have a [] but it must have the word Array as the suffix go clarify that T is of type Array.

**Comment by dgreene1** [comment]
@Asterix112 / @lazylazyllama the fix for this was released in 2.4.10. Thank you for informing us of the issue. Please let us know if it's resolved. You can also see what else was released in our release notes: https://github.com/lukeautry/tsoa/releases

Lastly, if you're enjoying using tsoa, then please consider sharing your enjoyment on Twitter, Reddit, or [dev.to](https://dev.to). :)

**Comment by Asterix112** [comment]
@dgreene1 @WoH Thank you for your fast work and the well done fix! Everything works as expected now.
